src/lapack/backends/rocsolver/rocsolver_scope_handle.cpp

```diff
@@ -0,0 +1,141 @@
+        // No context is installed and the suggested context is primary
+        // This is the most common case. We can activate the context in the
+        // thread and leave it there until all the PI context referring to the
+        // same underlying rocblas primary context are destroyed. This emulates
+        // the behaviour of the rocblas runtime api, and avoids costly context
+        // switches. No action is required on this side of the if.
+        needToRecover_ = !(original_ == nullptr);
+    }
+}
+
+RocsolverScopedContextHandler::~RocsolverScopedContextHandler() noexcept(false) {
+    if (needToRecover_) {
+        hipError_t err;
+        HIP_ERROR_FUNC(hipCtxSetCurrent, err, original_);
+    }
+    delete placedContext_;
+}
+
+void ContextCallback(void *userData) {
+    auto *ptr = static_cast<std::atomic<rocblas_handle> *>(userData);
+    if (!ptr) {
+        return;
+    }
+    auto handle = ptr->exchange(nullptr);
+    if (handle != nullptr) {
+        rocblas_status err1;
+        ROCSOLVER_ERROR_FUNC(rocblas_destroy_handle, err1, handle);
+        handle = nullptr;
+    }
+    else {
+        // if the handle is nullptr it means the handle was already destroyed by
+        // the rocblas_handle destructor and we're free to delete the atomic
+        // object.
+        delete ptr;
+    }
+}
+
+rocblas_handle RocsolverScopedContextHandler::get_handle(const sycl::queue &queue) {
+    auto piPlacedContext_ = reinterpret_cast<pi_context>(
+        sycl::get_native<sycl::backend::ext_oneapi_hip>(*placedContext_));
+    hipStream_t streamId = get_stream(queue);
+    rocblas_status err;
+    auto it = handle_helper.rocsolver_handle_mapper_.find(piPlacedContext_);
+    if (it != handle_helper.rocsolver_handle_mapper_.end()) {
+        if (it->second == nullptr) {
+            handle_helper.rocsolver_handle_mapper_.erase(it);
+        }
+        else {
+            auto handle = it->second->load();
+            if (handle != nullptr) {
+                hipStream_t currentStreamId;
+                ROCSOLVER_ERROR_FUNC(rocblas_get_stream, err, handle, &currentStreamId);
+                if (currentStreamId != streamId) {
+                    ROCSOLVER_ERROR_FUNC(rocblas_set_stream, err, handle, streamId);
+                }
+                return handle;
+            }
+            else {
+                handle_helper.rocsolver_handle_mapper_.erase(it);
+            }
+        }
+    }
+
+    rocblas_handle handle;
+
+    ROCSOLVER_ERROR_FUNC(rocblas_create_handle, err, &handle);
+    ROCSOLVER_ERROR_FUNC(rocblas_set_stream, err, handle, streamId);
+
+    auto insert_iter = handle_helper.rocsolver_handle_mapper_.insert(
+        std::make_pair(piPlacedContext_, new std::atomic<rocblas_handle>(handle)));
+
+    sycl::detail::pi::contextSetExtendedDeleter(*placedContext_, ContextCallback,
+                                                insert_iter.first->second);
+
+    return handle;
+}
```

Comment:
Just checking if the above instances of `rocblas` should be `rocsolver`

Reply:
Usage of rocsolver_handle is [deprecated](https://rocsolver.readthedocs.io/en/rocm-5.2.1/api_deprecated.html?highlight=rocsolver_handle#rocsolver-handle:~:text=Deprecated%20since%20version%203.5%3A%20Use%20rocblas_handle)